panic: interface conversion: interface is nil, not float64

Hi 

I'm using Kapacitor 0.12.0,
I recently started to get panic errors with Kapacitor:

```
panic: interface conversion: interface is nil, not float64

goroutine 296 [running]:
panic(0xc19100, 0xc827ebe340)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/influxdata/kapacitor.(*floatPointBulkAggregator).AggregateBatch(0xc829048480, 0xc8211e2000)
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/influxql.gen.go:78 +0x23b
github.com/influxdata/kapacitor.(*InfluxQLNode).runBatchInfluxQL(0xc8202d70e0, 0x0, 0x0)
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/influxql.go:137 +0x458
github.com/influxdata/kapacitor.(*InfluxQLNode).runInfluxQLs(0xc8202d70e0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/influxql.go:37 +0x83
github.com/influxdata/kapacitor.(*InfluxQLNode).(github.com/influxdata/kapacitor.runInfluxQLs)-fm(0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/influxql.go:28 +0x4c
github.com/influxdata/kapacitor.(*node).start.func1(0xc8202d70e0, 0x0, 0x0, 0x0)
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/node.go:128 +0x7f
created by github.com/influxdata/kapacitor.(*node).start
	/tmp/kapacitor/src/github.com/influxdata/kapacitor/node.go:129 +0x53
```

where afterwards Kapacitor stops.

Any idea why is it happening?

Nitz

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -58,6 +58,7 @@ For example, let's say we want to store all data that triggered an alert in Infl
 
 ### Bugfixes
 
+- [#499](https://github.com/influxdata/kapacitor/issues/499): Fix panic in InfluxQL nodes if field is missing or incorrect type.
 - [#441](https://github.com/influxdata/kapacitor/issues/441): Fix panic in UDF code.
 - [#429](https://github.com/influxdata/kapacitor/issues/429): BREAKING: Change TICKscript parser to be left-associative on equal precedence operators. For example previously this statement `(1+2-3*4/5)` was evaluated as `(1+(2-(3*(4/5))))`
     which is not the typical/expected behavior. Now using left-associative parsing the statement is evaluated as `((1+2)-((3*4)/5))`.

--- a/influxql.gen.go
+++ b/influxql.gen.go
@@ -33,13 +33,21 @@ func floatPopulateAuxFieldsAndTags(ap *influxql.FloatPoint, fieldsAndTags []stri
 	}
 }
 
-func (a *floatPointAggregator) AggregateBatch(b *models.Batch) {
+func (a *floatPointAggregator) AggregateBatch(b *models.Batch) error {
 	for _, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.(float64)
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp float64", a.field, value)
+		}
 		ap := &influxql.FloatPoint{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].(float64),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -52,14 +60,23 @@ func (a *floatPointAggregator) AggregateBatch(b *models.Batch) {
 
 		a.aggregator.AggregateFloat(ap)
 	}
+	return nil
 }
 
-func (a *floatPointAggregator) AggregatePoint(p *models.Point) {
+func (a *floatPointAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.(float64)
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp float64", a.field, value)
+	}
 	ap := &influxql.FloatPoint{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].(float64),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -71,6 +88,7 @@ func (a *floatPointAggregator) AggregatePoint(p *models.Point) {
 	}
 
 	a.aggregator.AggregateFloat(ap)
+	return nil
 }
 
 type floatPointBulkAggregator struct {
@@ -80,14 +98,22 @@ type floatPointBulkAggregator struct {
 	aggregator       pipeline.FloatBulkPointAggregator
 }
 
-func (a *floatPointBulkAggregator) AggregateBatch(b *models.Batch) {
+func (a *floatPointBulkAggregator) AggregateBatch(b *models.Batch) error {
 	slice := make([]influxql.FloatPoint, len(b.Points))
 	for i, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.(float64)
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp float64", a.field, value)
+		}
 		slice[i] = influxql.FloatPoint{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].(float64),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -99,14 +125,23 @@ func (a *floatPointBulkAggregator) AggregateBatch(b *models.Batch) {
 		}
 	}
 	a.aggregator.AggregateFloatBulk(slice)
+	return nil
 }
 
-func (a *floatPointBulkAggregator) AggregatePoint(p *models.Point) {
+func (a *floatPointBulkAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.(float64)
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp float64", a.field, value)
+	}
 	ap := &influxql.FloatPoint{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].(float64),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -118,6 +153,7 @@ func (a *floatPointBulkAggregator) AggregatePoint(p *models.Point) {
 	}
 
 	a.aggregator.AggregateFloat(ap)
+	return nil
 }
 
 type floatPointEmitter struct {
@@ -205,13 +241,21 @@ func integerPopulateAuxFieldsAndTags(ap *influxql.IntegerPoint, fieldsAndTags []
 	}
 }
 
-func (a *integerPointAggregator) AggregateBatch(b *models.Batch) {
+func (a *integerPointAggregator) AggregateBatch(b *models.Batch) error {
 	for _, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.(int64)
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp int64", a.field, value)
+		}
 		ap := &influxql.IntegerPoint{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].(int64),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -224,14 +268,23 @@ func (a *integerPointAggregator) AggregateBatch(b *models.Batch) {
 
 		a.aggregator.AggregateInteger(ap)
 	}
+	return nil
 }
 
-func (a *integerPointAggregator) AggregatePoint(p *models.Point) {
+func (a *integerPointAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.(int64)
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp int64", a.field, value)
+	}
 	ap := &influxql.IntegerPoint{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].(int64),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -243,6 +296,7 @@ func (a *integerPointAggregator) AggregatePoint(p *models.Point) {
 	}
 
 	a.aggregator.AggregateInteger(ap)
+	return nil
 }
 
 type integerPointBulkAggregator struct {
@@ -252,14 +306,22 @@ type integerPointBulkAggregator struct {
 	aggregator       pipeline.IntegerBulkPointAggregator
 }
 
-func (a *integerPointBulkAggregator) AggregateBatch(b *models.Batch) {
+func (a *integerPointBulkAggregator) AggregateBatch(b *models.Batch) error {
 	slice := make([]influxql.IntegerPoint, len(b.Points))
 	for i, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.(int64)
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp int64", a.field, value)
+		}
 		slice[i] = influxql.IntegerPoint{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].(int64),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -271,14 +333,23 @@ func (a *integerPointBulkAggregator) AggregateBatch(b *models.Batch) {
 		}
 	}
 	a.aggregator.AggregateIntegerBulk(slice)
+	return nil
 }
 
-func (a *integerPointBulkAggregator) AggregatePoint(p *models.Point) {
+func (a *integerPointBulkAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.(int64)
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp int64", a.field, value)
+	}
 	ap := &influxql.IntegerPoint{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].(int64),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -290,6 +361,7 @@ func (a *integerPointBulkAggregator) AggregatePoint(p *models.Point) {
 	}
 
 	a.aggregator.AggregateInteger(ap)
+	return nil
 }
 
 type integerPointEmitter struct {

--- a/influxql.gen.go.tmpl
+++ b/influxql.gen.go.tmpl
@@ -31,13 +31,21 @@ func {{.name}}PopulateAuxFieldsAndTags(ap *influxql.{{.Name}}Point, fieldsAndTag
 	}
 }
 
-func (a *{{.name}}PointAggregator) AggregateBatch(b *models.Batch) {
+func (a *{{.name}}PointAggregator) AggregateBatch(b *models.Batch) error {
 	for _, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.({{.Type}})
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp {{.Type}}", a.field, value)
+		}
 		ap := &influxql.{{.Name}}Point{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].({{.Type}}),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -50,14 +58,23 @@ func (a *{{.name}}PointAggregator) AggregateBatch(b *models.Batch) {
 		
 		a.aggregator.Aggregate{{.Name}}(ap)
 	}
+	return nil
 }
 
-func (a *{{.name}}PointAggregator) AggregatePoint(p *models.Point) {
+func (a *{{.name}}PointAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.({{.Type}})
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp {{.Type}}", a.field, value)
+	}
 	ap := &influxql.{{.Name}}Point{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].({{.Type}}),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -69,6 +86,7 @@ func (a *{{.name}}PointAggregator) AggregatePoint(p *models.Point) {
 	}
 	
 	a.aggregator.Aggregate{{.Name}}(ap)
+	return nil
 }
 
 
@@ -80,14 +98,22 @@ type {{.name}}PointBulkAggregator struct {
 	aggregator    pipeline.{{.Name}}BulkPointAggregator
 }
 
-func (a *{{.name}}PointBulkAggregator) AggregateBatch(b *models.Batch) {
+func (a *{{.name}}PointBulkAggregator) AggregateBatch(b *models.Batch) error {
 	slice := make([]influxql.{{.Name}}Point, len(b.Points))
 	for i, p := range b.Points {
+		value, ok := p.Fields[a.field]
+		if !ok {
+			return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+		}
+		typed, ok := value.({{.Type}})
+		if !ok {
+			return fmt.Errorf("field %s has wrong type: got %T exp {{.Type}}", a.field, value)
+		}
 		slice[i] = influxql.{{.Name}}Point{
 			Name:  b.Name,
 			Tags:  influxql.NewTags(p.Tags),
 			Time:  p.Time.UnixNano(),
-			Value: p.Fields[a.field].({{.Type}}),
+			Value: typed,
 		}
 		if a.topBottomInfo != nil {
 			// We need to populate the Aux fields
@@ -99,14 +125,23 @@ func (a *{{.name}}PointBulkAggregator) AggregateBatch(b *models.Batch) {
 		}
 	}
 	a.aggregator.Aggregate{{.Name}}Bulk(slice)
+	return nil
 }
 
-func (a *{{.name}}PointBulkAggregator) AggregatePoint(p *models.Point) {
+func (a *{{.name}}PointBulkAggregator) AggregatePoint(p *models.Point) error {
+	value, ok := p.Fields[a.field]
+	if !ok {
+		return fmt.Errorf("field %s missing from point cannot aggregate", a.field)
+	}
+	typed, ok := value.({{.Type}})
+	if !ok {
+		return fmt.Errorf("field %s has wrong type: got %T exp {{.Type}}", a.field, value)
+	}
 	ap := &influxql.{{.Name}}Point{
 		Name:  p.Name,
 		Tags:  influxql.NewTags(p.Tags),
 		Time:  p.Time.UnixNano(),
-		Value: p.Fields[a.field].({{.Type}}),
+		Value: typed,
 	}
 	if a.topBottomInfo != nil {
 		// We need to populate the Aux fields
@@ -118,6 +153,7 @@ func (a *{{.name}}PointBulkAggregator) AggregatePoint(p *models.Point) {
 	}
 
 	a.aggregator.Aggregate{{.Name}}(ap)
+	return nil
 }
 
 type {{.name}}PointEmitter struct {

--- a/influxql.go
+++ b/influxql.go
@@ -46,8 +46,8 @@ func (n *InfluxQLNode) runInfluxQLs([]byte) error {
 }
 
 type reduceContext interface {
-	AggregatePoint(p *models.Point)
-	AggregateBatch(b *models.Batch)
+	AggregatePoint(p *models.Point) error
+	AggregateBatch(b *models.Batch) error
 	EmitPoint() (models.Point, error)
 	EmitBatch() models.Batch
 	Time() time.Time
@@ -97,22 +97,28 @@ func (n *InfluxQLNode) runStreamInfluxQL() error {
 
 		}
 		if n.isStreamTransformation {
-			context.AggregatePoint(&p)
+			err := context.AggregatePoint(&p)
+			if err != nil {
+				n.logger.Println("E! failed to aggregate point:", err)
+			}
 			p, ok = n.ins[0].NextPoint()
 
-			err := n.emit(context)
+			err = n.emit(context)
 			if err != nil && err != ErrEmptyEmit {
-				return err
+				n.logger.Println("E! failed to emit stream:", err)
 			}
 		} else {
 			if p.Time.Equal(context.Time()) {
-				context.AggregatePoint(&p)
+				err := context.AggregatePoint(&p)
+				if err != nil {
+					n.logger.Println("E! failed to aggregate point:", err)
+				}
 				// advance to next point
 				p, ok = n.ins[0].NextPoint()
 			} else {
 				err := n.emit(context)
 				if err != nil {
-					return err
+					n.logger.Println("E! failed to emit stream:", err)
 				}
 
 				// Nil out reduced point
@@ -149,10 +155,13 @@ func (n *InfluxQLNode) runBatchInfluxQL() error {
 		}
 
 		context := createFn(c)
-		context.AggregateBatch(&b)
+		err = context.AggregateBatch(&b)
+		if err != nil {
+			n.logger.Println("E! failed to aggregate batch:", err)
+		}
 		err = n.emit(context)
 		if err != nil {
-			return err
+			n.logger.Println("E! failed to emit batch:", err)
 		}
 	}
 	return nil

